File: sdk/npm/src/providers/temp-mail-io.ts
import { InternalEmailInfo, Email, Channel } from '../types';
import { normalizeEmail } from '../normalize';
import { fetchWithTimeout } from '../retry';

const CHANNEL: Channel = 'temp-mail-io';
const BASE_URL = 'https://api.internal.temp-mail.io/api/v3';
const PAGE_URL = 'https://temp-mail.io/en';

let cachedCorsHeader: string | null = null;

async function fetchCorsHeader(): Promise<string> {
  if (cachedCorsHeader) return cachedCorsHeader;

  try {
    const response = await fetchWithTimeout(PAGE_URL, {
      headers: {
        'User-Agent':
          'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/144.0.0.0 Safari/537.36',
      },
    });
    const html = await response.text();
    const match = html.match(/mobileTestingHeader\s*:\s*"([^"]+)"/);
    if (match) {
      cachedCorsHeader = match[1];
      return cachedCorsHeader;
    }
  } catch {
    /* fallback below */
  }

  cachedCorsHeader = '1';
  return cachedCorsHeader;
}

async function getApiHeaders(): Promise<Record<string, string>> {
  const corsHeader = await fetchCorsHeader();
  return {
    'Content-Type': 'application/json',
    'Application-Name': 'web',
    'Application-Version': '4.0.0',
    'X-CORS-Header': corsHeader,
    'User-Agent':
      'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/144.0.0.0 Safari/537.36 Edg/144.0.0.0',
    origin: 'https://temp-mail.io',
    referer: 'https://temp-mail.io/',
  };
}

export async function generateEmail(): Promise<InternalEmailInfo> {
  const headers = await getApiHeaders();
  const response = await fetchWithTimeout(`${BASE_URL}/email/new`, {
    method: 'POST',
    headers,
    body: JSON.stringify({ min_name_length: 10, max_name_length: 10 }),
  });

  if (!response.ok) {
    throw new Error(`temp-mail-io generate: ${response.status}`);
  }

  const data = await response.json();
  if (!data.email || !data.token) {
    throw new Error('temp-mail-io: invalid generate response');
  }

  return {
    channel: CHANNEL,
    email: data.email,
    token: data.token,
  };
}

export async function getEmails(email: string): Promise<Email[]> {
  const headers = await getApiHeaders();
  const response = await fetchWithTimeout(`${BASE_URL}/email/${email}/messages`, {
    method: 'GET',
    headers,
  });

  if (!response.ok) {
    throw new Error(`temp-mail-io messages: ${response.status}`);
  }

  const data = await response.json();
  const rawEmails = Array.isArray(data) ? data : [];
  return rawEmails.map((raw: any) => normalizeEmail(raw, email));
}

File: sdk/go/provider/temp_mail_io.go
package provider

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"regexp"
	"sync"

	http "github.com/bogdanfinn/fhttp"
)

const tempMailIoBaseURL = "https://api.internal.temp-mail.io/api/v3"
const tempMailIoPageURL = "https://temp-mail.io/en"

var (
	tempMailIoHeaderMu     sync.Mutex
	tempMailIoCachedHeader string
)

type tempMailIoGenerateResponse struct {
	Email string `json:"email"`
	Token string `json:"token"`
}

func tempMailIoFetchCorsHeader() string {
	tempMailIoHeaderMu.Lock()
	defer tempMailIoHeaderMu.Unlock()

	if tempMailIoCachedHeader != "" {
		return tempMailIoCachedHeader
	}

	req, err := http.NewRequest("GET", tempMailIoPageURL, nil)
	if err == nil {
		req.Header.Set("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/144.0.0.0 Safari/537.36")
		resp, reqErr := HTTPClient().Do(req)
		if reqErr == nil && resp != nil {
			body, readErr := io.ReadAll(resp.Body)
			resp.Body.Close()
			if readErr == nil {
				re := regexp.MustCompile(`mobileTestingHeader\s*:\s*"([^"]+)"`)
				if match := re.FindSubmatch(body); len(match) == 2 {
					tempMailIoCachedHeader = string(match[1])
					return tempMailIoCachedHeader
				}
			}
		} else if resp != nil {
			resp.Body.Close()
		}
	}

	tempMailIoCachedHeader = "1"
	return tempMailIoCachedHeader
}

func tempMailIoApplyHeaders(req *http.Request) {
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Application-Name", "web")
	req.Header.Set("Application-Version", "4.0.0")
	req.Header.Set("X-CORS-Header", tempMailIoFetchCorsHeader())
	req.Header.Set("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/144.0.0.0 Safari/537.36 Edg/144.0.0.0")
	req.Header.Set("Origin", "https://temp-mail.io")
	req.Header.Set("Referer", "https://temp-mail.io/")
}

func TempMailIoGenerate() (*CreatedMailbox, error) {
	reqBody, _ := json.Marshal(map[string]int{
		"min_name_length": 10,
		"max_name_length": 10,
	})

	req, err := http.NewRequest("POST", tempMailIoBaseURL+"/email/new", bytes.NewReader(reqBody))
	if err != nil {
		return nil, err
	}
	tempMailIoApplyHeaders(req)

	resp, err := HTTPClient().Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("temp-mail-io generate: %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var data tempMailIoGenerateResponse
	if err := json.Unmarshal(body, &data); err != nil {
		return nil, err
	}
	if data.Email == "" || data.Token == "" {
		return nil, fmt.Errorf("temp-mail-io: invalid generate response")
	}

	return &CreatedMailbox{Channel: "temp-mail-io", Email: data.Email, Token: data.Token}, nil
}

func TempMailIoGetEmails(email string) ([]NormEmail, error) {
	req, err := http.NewRequest("GET", tempMailIoBaseURL+"/email/"+email+"/messages", nil)
	if err != nil {
		return nil, err
	}
	tempMailIoApplyHeaders(req)

	resp, err := HTTPClient().Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("temp-mail-io messages: %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var rows []map[string]interface{}
	if err := json.Unmarshal(body, &rows); err != nil {
		return nil, err
	}

	emails := make([]NormEmail, 0, len(rows))
	for _, raw := range rows {
		emails = append(emails, NormalizeMap(raw, email))
	}
	return emails, nil
}
